src/Servers/Kestrel/Transport.MsQuic/src/Internal/QuicSession.cs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using static Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal sealed class QuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _registration;
+
+        internal QuicSession(MsQuicApi registration, IntPtr nativeObjPtr)
+        {
+            _registration = registration;
+            _nativeObjPtr = nativeObjPtr;
+        }
+
+        public async ValueTask<MsQuicConnection> ConnectionOpenAsync(IPEndPoint endpoint, MsQuicTransportContext context)
```

Comment:
Open and Start are explicitly separate in the msquic API for a reason. For the 0-RTT case, you open streams and send on them **before** starting the connection.